File: Homework/Hw2/Hw2.py
class Hw2:
    def __init__(self, first_bad_version):
        self.first_bad_version = first_bad_version
    
    def firstBadVersion(self, n):
        left = 1
        right = n
        
        while left < right:
            mid = left + (right - left) // 2
            if self.isBadVersion(mid):
                right = mid
            else:
                left = mid + 1
                
        return left

    def isBadVersion(self, version):
        return version >= self.first_bad_version

# Example usage
if __name__ == "__main__":
    solution = Hw2(4)  # Assuming version 4 is the first bad version
    print(solution.firstBadVersion(5))  # Should print 4

File: Homework/Hw2/Hw2.java
public class Hw2 {
    // Assuming a method isBadVersion(int version) exists and returns a boolean

    // Method to find the first bad version
    public int firstBadVersion(int n) {
        int left = 1;
        int right = n;
        
        while (left < right) {
            int mid = left + (right - left) / 2;
            if (isBadVersion(mid)) {
                right = mid;
            } else {
                left = mid + 1;
            }
        }
        
        return left;
    }
    
    // Dummy isBadVersion method (replace or implement accordingly)
    boolean isBadVersion(int version) {
        // Implement your mechanism to identify if a version is bad
        // For example, return version >= firstBadVersion;
        return false; // Placeholder implementation
    }
}
